src/main/ruby/truffleruby/core/truffle/polyglot.rb

```diff
@@ -312,6 +312,14 @@ def backtrace_locations
     def backtrace
       backtrace_locations&.map(&:to_s)
     end
+
+    def marshal_dump
+      raise TypeError, "TruffleRuby foreign exceptions cannot be dumped: #{message}"
+    end
+
+    def marshal_load(...)
+      raise TypeError, 'TruffleRuby foreign exceptions cannot be restored'
```

Comment:
```suggestion
      raise TypeError, 'Foreign exception cannot be restored'
```

Reply:
I think making it clear it's a TruffleRuby limitation is more helpful. Just saying "foreign exception" could imply the exception comes from a native extension and would affect MRI as well.

Reply:
If CRuby would see e.g. a C++ exception I'm pretty sure they wouldn't be able to serialize it either without explicit support.
I think it's expected exceptions can be different when running on a different Ruby implementation, and people should know which Ruby implementation they are using, so the `TruffleRuby` feels redundant to me.

Reply:
A problem with including TruffleRuby there is it sounds like this is a "TruffleRuby exception" but actually it's an exception "not from TruffleRuby but from some other GraalVM language".
We have the same problem with `TruffleRuby doesn't have a case for the ` for `UnsupportedSpecializationException` and it feels wrong to me as well.

Reply:
If someone is porting an application that's been running on MRI to TruffleRuby, it's helpful to know when an exception is due to a difference or limitation of the new runtime. I work on the runtime and it'd be helpful to know when I'm dealing with a polyglot foreign exception and not some other type of foreign exception. We're changing the exception type, so that can't be relied upon.